docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,635 @@
+      to an epoch at UTC midnight on January 1, 1970.
+- decimal: exact decimal value represented as an integer value in two's complement.
+- binary: an variable-length array of bytes.
+- array type: only allow numeric component. Other arrays will be taken as List. The implementation should support the
+  interoperability between array and list.
+    - array: multidimensional array which every sub-array can have different size but all have same type.
+    - bool_array: one dimension int16 array.
+    - int16_array: one dimension int16 array.
+    - int32_array: one dimension int32 array.
+    - int64_array: one dimension int64 array.
+    - float16_array: one dimension half_float_16 array.
+    - float32_array: one dimension float32 array.
+    - float64_array: one dimension float64 array.
+- tensor: a multidimensional dense array of fixed-size values such as a NumPy ndarray.
+- sparse tensor: a multidimensional array whose elements are almost all zeros.
+- arrow record batch: an arrow [record batch](https://arrow.apache.org/docs/cpp/tables.html#record-batches) object.
+- arrow table: an arrow [table](https://arrow.apache.org/docs/cpp/tables.html#tables) object.
+
+### Type disambiguation
+
+Due to differences between type systems of languages, those types can't be mapped one-to-one between languages. When
+deserializing, Fury use the target data structure type and the data type in the data jointly to determine how to
+deserialize and populate the target data structure. For example:
+
+```java
+class Foo {
+  int[] intArray;
+  Object[] objects;
+  List<Object> objectList;
+}
+
+class Foo2 {
+  int[] intArray;
+  List<Object> objects;
+  List<Object> objectList;
+}
+```
+
+`intArray` has `int32_array` type. But both `objects` and `objectList` field in the serialize data have `list` data
+type. When deserializing, the implementation will create an `Object` array for `objects`, but create a `ArrayList`
+for `objectList` to populate it's elements. And the serialized data of `Foo` can be deserialized into `Foo2` too.
+
+Users can also provide meta hint for fields of a type, or the type whole. Here is an example in java which use
+annotation to provide such information.
+
+```java
+
+@TypeInfo(fieldsNullable = false, trackingRef = false, polymorphic = false)
+class Foo {
+  @FieldInfo(trackingRef = false)
+  int[] intArray;
+  @FieldInfo(polymorphic = true)
+  Object object;
+  @FieldInfo(tagId = 1, nullable = true)
+  List<Object> objectList;
+}
+```
+
+Such information can be provided in other languages too:
+
+- cpp: use macro and template.
+- golang: use struct tag.
+- python: use typehint.
+- rust: use macro.
+
+### Type ID
+
+All internal data types are expressed using a ID in range `-64~-1`. Users can use `0~32703` for representing their
+types. At runtime, all type ids are added by `64`, and then encoded as an unsigned varint.
```

Comment:
`unsigned varint`

So data type is different from encoding, right? There is no unsigned type, and also varint type.

Reply:
@PragmaTwice I added `var_int32/fixed_int32/fixed_int64/var_int64/sli_int64` to type systems, user can use type alias ot annotation to mark such types:
```java
class Foo {
  List<@VarInt64 Integer> list;
}
```
```golang
type varint64 int64

type Foo struct {
  []varint64 list
}
```

Reply:
Unsigned types not added, since they are not supported in every language, take `int64` in java as an example.